core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.executor
+
+import java.io._
+import java.nio.charset.Charset
+import java.nio.file.{Files, Paths}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.mutable.Queue
+
+import org.apache.spark.internal.Logging
+
+
+// Some of the ideas here are taken from the ProcfsBasedProcessTree class in hadoop
+// project.
+class ProcfsBasedSystems  extends ProcessTreeMetrics with Logging {
+  val procfsDir = "/proc/"
+  var isAvailable: Boolean = isItProcfsBased
+  val pid: Int = computePid()
+  val ptree: scala.collection.mutable.Map[ Int, Set[Int]] =
+    scala.collection.mutable.Map[ Int, Set[Int]]()
+  val PROCFS_STAT_FILE = "stat"
+
+  createProcessTree
+
+  def isItProcfsBased: Boolean = {
+    val testing = sys.env.contains("SPARK_TESTING") || sys.props.contains("spark.testing")
+    if (testing) {
+      return true
+    }
+    try {
+      if (!Files.exists(Paths.get(procfsDir))) {
+        return false
+      }
+    }
+    catch {
+      case f: FileNotFoundException => return false
+    }
+    true
+  }
+
+
+  def computePid(): Int = {
+    if (!isAvailable) {
+      return -1;
+    }
+    try {
+      val cmd = Array("bash", "-c", "echo $PPID")
```

Comment:
you might want to at least drop a comment in here that this can be simplified in java9:
https://docs.oracle.com/javase/9/docs/api/java/lang/ProcessHandle.html

Reply:
will add a comment.